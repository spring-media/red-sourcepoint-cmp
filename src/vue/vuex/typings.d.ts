import { CustomPurpose, CustomVendor } from '../../sourcepoint/typings';

export type SourcepointModuleState = {
  consentedCustomVendors: CustomVendor[];
  consentedCustomPurposes: CustomPurpose[];
  grantedVendors: string[];
  consentReady: boolean;
};
