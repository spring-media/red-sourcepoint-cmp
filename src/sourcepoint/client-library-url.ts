export const getClientLibraryUrl = (): string => `https://cdn.privacy-mgmt.com/wrapperMessagingWithoutDetection.js`;
