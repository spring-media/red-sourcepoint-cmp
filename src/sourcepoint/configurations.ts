import { ConfigurationParameters, SourcepointConfigurationObject } from './typings';

export const createConfiguration = (parameters: ConfigurationParameters): Partial<SourcepointConfigurationObject> => ({
  config: {
    accountId: parameters.accountId,
    wrapperAPIOrigin: 'https://wrapper-api.sp-prod.net/tcfv2',
    propertyId: parameters.propertyId,
    mmsDomain: 'https://message.sp-prod.net',
    events: parameters.events,
  },
});
