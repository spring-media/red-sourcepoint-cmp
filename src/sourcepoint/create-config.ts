import { Config } from './typings';

export const createConfig = (parameters: Config): Config => ({
  accountId: 0,
  wrapperAPIOrigin: 'https://wrapper-api.sp-prod.net/tcfv2',
  propertyId: 0,
  mmsDomain: 'https://message.sp-prod.net',
  ...parameters,
});
