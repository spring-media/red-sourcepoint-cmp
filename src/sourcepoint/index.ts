export * from './iab-stub-script';
export * from './client-library-url';
export * from './has-custom-consent';
export * from './privacy-manager';
export * from './create-config';
export * from './get-custom-vendor-consents';
