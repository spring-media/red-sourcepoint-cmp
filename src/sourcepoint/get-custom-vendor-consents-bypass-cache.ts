import { CustomVendorConsentsResult } from './typings';
import { getIABStubScript } from './iab-stub-script';
import { getClientLibraryUrl } from './client-library-url';
import { getCustomVendorConsents } from './get-custom-vendor-consents';

export const getCustomVendorConsentsBypassCache = (): Promise<CustomVendorConsentsResult> => {
  // Check browser support for srcdoc attribute (i'm talking to you IE)
  if (!Boolean('srcdoc' in document.createElement('iframe'))) {
    return getCustomVendorConsents();
  }

  if (!window._sp_?.config) {
    return getCustomVendorConsents();
  }

  const { accountId, wrapperAPIOrigin, propertyId, mmsDomain } = window._sp_.config;

  return new Promise((resolve) => {
    const iframe = document.createElement('iframe');
    iframe.srcdoc = `
      <script>${getIABStubScript()}</script>
      <script>
        window._sp_ = {
          config: {
            accountId: ${accountId},
            wrapperAPIOrigin: '${wrapperAPIOrigin}',
            propertyId: ${propertyId},
            mmsDomain: '${mmsDomain}',
          },
        };
      </script>
      <script src="${getClientLibraryUrl()}"></script>
      <script>
        __tcfapi('getCustomVendorConsents', 2, function(consents) {
          window.parent.postMessage({ type: 'cmp', consents: consents }, '*');
        });
      </script>
    `;

    const onMessage = (event: MessageEvent): void => {
      if (!event.data || event.data.type !== 'cmp') {
        return;
      }

      window.removeEventListener('message', onMessage);
      document.body.removeChild(iframe);

      return resolve(event.data.consents);
    };

    window.addEventListener('message', onMessage, false);

    document.body.appendChild(iframe);
  });
};
