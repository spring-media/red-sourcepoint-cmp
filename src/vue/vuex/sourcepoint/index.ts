import { Commit, Dispatch, Module } from 'vuex';
import { getCustomVendorConsents } from '../../../sourcepoint';
import { SourcepointModuleState } from '../typings';
import { CustomPurpose, CustomVendor } from '../../../sourcepoint/typings';
import {
  configureGrants,
  configureVendorPurposeMapping,
  getGrantedVendors,
  loadVendorPurposeMapping,
} from '../../../vendor-mapping';

const moduleState: SourcepointModuleState = {
  consentedCustomVendors: [],
  consentedCustomPurposes: [],
  grantedVendors: [],
};

export const mutations = {
  setCustomVendorConsents(state: SourcepointModuleState, payload: CustomVendor[]): void {
    state.consentedCustomVendors = payload;
  },
  setCustomPurposeConsents(state: SourcepointModuleState, payload: CustomPurpose[]): void {
    state.consentedCustomPurposes = payload;
  },
  setGrantedVendors(state: SourcepointModuleState, payload: string[]): void {
    state.grantedVendors = payload;
  },
};

export const getters = {
  vendorHasConsent: (state: SourcepointModuleState) => (vendorId: string): boolean =>
    state.grantedVendors.includes(vendorId),
};

export const actions = {
  async refreshCustomVendorConsents({ commit }: { commit: Commit }): Promise<void> {
    const { consentedPurposes = [], consentedVendors = [], grants = {} } = await getCustomVendorConsents();

    configureGrants(grants);

    commit('setCustomVendorConsents', consentedVendors);
    commit('setCustomPurposeConsents', consentedPurposes);
    commit('setGrantedVendors', getGrantedVendors());
  },
  async bootstrapConsents({ dispatch }: { dispatch: Dispatch }, payload: { propertyId: number }): Promise<void> {
    await dispatch('refreshCustomVendorConsents');

    const mappings = await loadVendorPurposeMapping(payload.propertyId);

    configureVendorPurposeMapping(mappings);
  },
};

export const sourcepoint: Module<SourcepointModuleState, Record<string, unknown>> = {
  namespaced: true,
  state: moduleState,
  mutations,
  actions,
  getters,
};
