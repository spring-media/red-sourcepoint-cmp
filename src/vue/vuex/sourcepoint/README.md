# sourcepoint Vuex Module

## Effects

### reloadPageOnReject

This effect reloads the page after a vendor or purpose has been rejected.

## Example

```javascript
import Vue from 'vue';
import Vuex from 'vuex';
import { sourcepoint } from '@spring-media/red-sourcepoint-cmp/dist/esm/vue/vuex/sourcepoint';
import { reloadPageOnReject } from '@spring-media/red-sourcepoint-cmp/dist/esm/vue/vuex/sourcepoint/effects';

Vue.use(Vuex);

const store = new Vuex.Store({
  strict: true,
  state: {},
  modules: {
    sourcepoint,
  },
});

reloadPageOnReject(store);
``` 
