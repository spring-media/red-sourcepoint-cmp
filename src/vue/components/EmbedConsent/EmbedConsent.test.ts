import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import { EmbedConsent } from './';
import { sourcepoint } from '../../vuex/sourcepoint';

const localVue = createLocalVue();

localVue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    sourcepoint,
  },
});

describe('EmbedConsent component', () => {
  afterEach(() => {
    store.commit('sourcepoint/setCustomVendorConsents', []);
    store.commit('sourcepoint/setCustomPurposeConsents', []);
  });

  it('should render the placeholder by default', () => {
    const wrapper = mount(EmbedConsent, {
      propsData: {
        customConsents: {},
        vendorId: 'abc',
        privacyManagerId: 123,
      },
      store,
      localVue,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render the embed in case of a consented vendor', async () => {
    store.commit('sourcepoint/setCustomVendorConsents', [{ _id: 'abc' }]);

    const wrapper = mount(EmbedConsent, {
      propsData: {
        customConsents: {},
        content: '<div>Embed Content</div>',
        vendorId: 'abc',
        privacyManagerId: 123,
      },
      store,
      localVue,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.element).toMatchInlineSnapshot(`
      <div>
        <div>
          Embed Content
        </div>
      </div>
    `);
  });
});
