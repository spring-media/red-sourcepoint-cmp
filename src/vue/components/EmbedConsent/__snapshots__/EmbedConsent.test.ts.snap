// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmbedConsent component should render the placeholder by default 1`] = `
<div
  class="embed-placeholder__container"
>
  <div
    class="embed-placeholder__header"
  >
    <svg
      fill="none"
      height="265"
      width="100%"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="#E9ECEF"
        height="8"
        width="81"
        x="38"
        y="62"
      />
       
      <rect
        fill="#E9ECEF"
        height="8"
        width="81"
        x="38"
        y="88"
      />
       
      <rect
        fill="#E9ECEF"
        height="8"
        width="61"
        x="38"
        y="75"
      />
       
      <rect
        fill="#E9ECEF"
        height="8"
        width="81"
        x="38"
        y="173"
      />
       
      <rect
        fill="#E9ECEF"
        height="8"
        width="81"
        x="38"
        y="199"
      />
       
      <rect
        fill="#E9ECEF"
        height="8"
        width="61"
        x="38"
        y="186"
      />
       
      <path
        d="M289.782 110.125L313 78L343 78"
        stroke="#868E96"
        stroke-width="3"
      />
       
      <path
        d="M287.287 170.75C292.847 161.12 296.117 150.339 296.845 139.242C297.572 128.145 295.737 117.03 291.481 106.756C287.226 96.4817 280.663 87.3243 272.303 79.9919C263.942 72.6596 254.006 67.3489 243.264 64.4707L239.511 78.4766C248.105 80.7792 256.053 85.0278 262.742 90.8936C269.431 96.7595 274.68 104.085 278.085 112.305C281.49 120.524 282.958 129.416 282.376 138.294C281.794 147.171 279.178 155.796 274.729 163.5L287.287 170.75Z"
        fill="#868E96"
      />
       
      <path
        d="M241.671 64.0628C230.542 61.3497 218.928 61.3124 207.781 63.954C196.635 66.5956 186.273 71.8409 177.545 79.26C168.817 86.6791 161.971 96.0608 157.568 106.636C153.166 117.212 151.332 128.68 152.217 140.101L166.673 138.981C165.965 129.844 167.433 120.669 170.955 112.209C174.477 103.749 179.953 96.2433 186.936 90.308C193.918 84.3727 202.208 80.1765 211.125 78.0632C220.042 75.9499 229.333 75.9797 238.237 78.1502L241.671 64.0628Z"
        fill="#CED4DA"
      />
       
      <path
        d="M152.396 142.061C153.976 157.128 160.236 171.321 170.296 182.647C180.357 193.974 193.713 201.864 208.489 205.209C223.264 208.555 238.716 207.188 252.675 201.301C266.633 195.414 278.397 185.302 286.314 172.386L273.951 164.808C267.618 175.142 258.207 183.231 247.04 187.941C235.873 192.651 223.511 193.744 211.691 191.067C199.871 188.391 189.186 182.079 181.137 173.018C173.088 163.957 168.08 152.602 166.816 140.549L152.396 142.061Z"
        fill="#CED4DA"
      />
       
      <rect
        fill="#868E96"
        height="8"
        width="81"
        x="349"
        y="62"
      />
       
      <rect
        fill="#868E96"
        height="8"
        width="81"
        x="349"
        y="88"
      />
       
      <rect
        fill="#868E96"
        height="8"
        width="61"
        x="349"
        y="75"
      />
    </svg>
  </div>
   
  <div
    class="embed-placeholder__headline"
  >
    
      An dieser Stelle findest du Inhalte von Drittanbietern
    
  </div>
   
  <div
    class="embed-placeholder__description"
  >
    
      Um mit Inhalten von Drittanbietern zu interagieren oder diese darzustellen, brauchen wir deine Zustimmung.
    
  </div>
   
  <button
    class="embed-placeholder__button"
  >
    
          externen Inhalt aktivieren
        
  </button>
   
  <div
    class="embed-placeholder__footer-text"
  >
    
        Ich bin damit einverstanden, dass mir Inhalte von Drittanbietern angezeigt werden. Damit können
        personenbezogene Daten an Drittanbieter übermittelt werden. Dazu ist ggf. die Speicherung von Cookies auf
        deinem Gerät notwendig. Mehr Informationen dazu findest du
        
    <a
      class="embed-placeholder__text-link"
      href="#"
      rel="noopener noreferrer"
      target="_blank"
    >
      hier
    </a>
    .
      
  </div>
</div>
`;
